### DISCUSSION TITLE
Density primitive/pressure primitive and entropy variable implementation in NS modulus

### DISCUSSION THREAD

**Discussion by Silverwing747** [idaholab/moose#29107] [Q&A Modules: Navier-Stokes] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Navier Stokes is the most appropriate category for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Question

Hello everyone,

I recently switched to use MOOSE, and I have been really enjoying the platform so far! 

I have also been play with NS Modulus for a while, and I think the modulus only allow you to use one set of variables to solve the problem. However, for my research, I need to implement different sets of variables to solve Navier-Stokes equations. In specific, I will need to solve:
1. Density primitive variable $\mathbf{Y} = [\rho,\mathbf{u},T]^T$
2. Pressure primitive variable $\mathbf{Y} = [p,\mathbf{u},T]^T$
3. Entropy variable $\mathbf{Y} = \frac{1}{T}[\mu,\mathbf{u},-1]^T$

where $\mu$ is chemical potential, $T$ is temperature, $p$ is pressure, $\rho$ is density and $\mathbf{u}$ is velocity. 

Since these funcions are not there, I will have to implement them myself. I think the easiest option is write three different Action files where each handles their own variable, but this choice lead to many code repeatition, and potentially difficult to maintain in the future. 

I am wondering if any one knows, or have any sugguestion, on how to design the code structure that:
1. Allows simple change of variables in the input file
2. Minimizes the code repeatition -- in principle, the main kernel should be same for all variables, and before enter the main kernel we convert the all variables into some standard variables. 
3. Keeps the code as similar to current NS module structure as possible. Also reuse as much as of the current NS module. 

Any feedbacks are welcome! 

### Additional information

Mesh size and type:
 Reynolds number:
 Discretization (finite element CG/DG, finite volume, etc):
 Models (turbulence, porous media, etc):
 Solver method (fully coupled, segregated, multiapps, etc):
 Base input you started from:

**Comment by Silverwing747** [comment]
Reference paper: "A comparative study of different sets of variables for solving compressible and incompressible flows" https://doi.org/10.1016/S0045-7825(97)00043-1

**Comment by GiudGiud** [comment, marked as answer]
Hello

I would focus on getting something working then you can worry about using inheritance and other techniques to reduce the code duplication. 

To do that, make a copy of every kernel you need, prefixing the new object name with "ConservedVar" for example, then modify them to use the conserved variables instead of the primitive ones

Note that both the thermal hydraulics module and the compressible finite volume kernels (CNSFV) already use the conservative variables.
